src/backends/caffe2/caffe2inputimg.cc

```diff
@@ -37,23 +45,34 @@ namespace dd
     _ids = _uris;
   }
 
+  void ImgCaffe2InputFileConn::transform_train(const APIData &) {
+    _shuffle = true;
+    //TODO
+  }
+
   int ImgCaffe2InputFileConn::get_tensor_test(caffe2::TensorCPU &tensor, int num) {
     if (!_images.size()) {
-      return 0;
+      return 0; // No more data
+    }
+    int w = _images[0].cols;
+    int h = _images[0].rows;
+    if (num < 0) {
+      num = _images.size(); // No fixed batch_size
     }
-    auto w = _images[0].cols;
-    auto h = _images[0].rows;
+
+    // Resize and prefill with 0s
     tensor.Resize(std::vector<caffe2::TIndex>({num, 3, h, w}));
-    auto data = tensor.mutable_data<float>();
+    float *data = tensor.mutable_data<float>();
     std::fill(data, data + tensor.size(), 0);
```

Comment:
I don't believe we need to pre-fill, the tensor should be the exact size already.